### PR TITLE
Expose window visibility state

### PR DESCRIPTION
**Pull request by xacrimon**
- [x] Tested on all platforms changed
- [x] Compilation warnings were addressed
- [x] `cargo fmt` has been run on this branch
- [x] `cargo doc` builds successfully
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [x] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by xacrimon** [comment]
This issue solves #2022 and provides a functionality for helping users avoid https://github.com/gfx-rs/wgpu/issues/1936

File: src/window.rs
Comment (maroider) [inline src/window.rs:778]:
```suggestion
    /// - **iOS / Android / Web / Windows / X11 / Wayland:** Unsupported and always returns `false`
```

File: src/window.rs
Comment (maroider) [inline src/window.rs:780]: I'm not sure if this is the best name. The only alternative I can think of is `occluded`, but that inverts the logic.

File: CHANGELOG.md
Comment (maroider) [inline CHANGELOG.md:17]:
```suggestion
- On macOS , expose `NSWindowOcclusionState` through `Window::visible`. This method returns `false` on all other platforms.
```

**Review by maroider — COMMENTED** [review]
cc @madsmtm

File: src/window.rs
Comment (madsmtm) [reply src/window.rs:778]: Also, the code returns `true`

File: src/window.rs
Comment (madsmtm) [inline src/window.rs:774]: Can you give an example of this? E.g. why is it not enough to just check the `visible` property (via. [`-isVisible`](https://developer.apple.com/documentation/appkit/nswindow/1419132-visible?language=objc))

**Review by madsmtm — CHANGES_REQUESTED** [review]
Can this be supported on other platforms? If not, maybe we should just expose it in `WindowExtMacOS` instead?

Edit: Ooops, sorry, missed the linked issue.

File: src/window.rs
Comment (madsmtm) [reply src/window.rs:780]: This is intended as a way to see whether we should render to the window, so maybe name it as such: `should_render`?

Comment (xacrimon) [reply src/window.rs:780]: Yeah that's what it is. We could name it occluded but that makes it confusing since it could be partially occluded or something else. `should_render` could work but it feels weird to name it after the purpose of which it exists and what we think the user will do with it rather than what it actually is. Describing the purpose of it via docs as I did feels like it makes more sense but I am happy to change it. I based the naming on how the rest of the api is named after what it is rather than what it's for. @madsmtm

File: src/window.rs
Comment (xacrimon) [reply src/window.rs]: fixed

File: CHANGELOG.md
Comment (xacrimon) [reply CHANGELOG.md]: fixed

File: src/window.rs
Comment (xacrimon) [reply src/window.rs:774]: @madsmtm `visible` returns true even when the window in question is fully occluded by other windows, i.e it's not hidden in the window manager. `AnimationCoreKit` on Mac as of some recent Big Sur versions on amd64 (can't confirm this) and the M1 chip stops periodical resource reclamation tasks as a battery saving measure when the window is hidden *or* fully occluded. As described in the linked `wgpu` issue this means that graphics applications may exhaust resources like memory and `MTLDrawable`'s when the window becomes fully occluded, causing an error within Metal.

**Comment by xacrimon** [comment]
@madsmtm I believe Windows also has an API about this since chrome relies on it to pause rendering (unsure about X11/Wayland) but I lack the knowledge and hardware to test it. That said this is not of immediate concern of any other platform than MacOS since all other platforms don't change their resource handling behaviour when the window is fully occluded.

File: src/window.rs
Comment (madsmtm) [reply src/window.rs:780]: I can see your reasoning, `visible` is fine to me

File: CHANGELOG.md
Comment (madsmtm) [inline CHANGELOG.md:17]:
```suggestion
- On macOS , expose `NSWindowOcclusionState` through `Window::visible`. This method returns `true` on all other platforms.
```

**Comment by madsmtm** [comment]
> I believe Windows also has an API about this

Then I think `Window::visible` is the correct place to put this.

**Comment by maroider** [comment]
I unfortunately don't have any knowledge of the Windows side of this, but I did find [this document](https://github.com/chromium/chromium/blob/master/docs/windows_native_window_occlusion_tracking.md) on the chromium GitHub page, which indicates that tracking occlusion on Windows is a lot more work.

**Comment by ArturKovacs** [comment]
@madsmtm I really think you should be memberized, especially because I'm considering moving away from being a maintainer.

File: CHANGELOG.md
Comment (xacrimon) [reply CHANGELOG.md]: fixed

**Comment by xacrimon** [comment]
@maroider @madsmtm fixed the last open review, lmk if I need to anything to move this forward

**Comment by madsmtm** [comment]
Bump @xacrimon

**Review by kchibisov — CHANGES_REQUESTED** [review]
I don't like the addition and it won't work on Wayland for sure, since your window doesn't know whether it's visible or not. It seems like the idea is to draw when window is visible, but each OS provides far better mechanism to do so and I'd suggest to leave that to downstream clients, since winit doesn't perform frame scheduling, etc, etc.

If someone rely on `visible` in drawing they'll end up always drawing on platforms like Wayland.

**Comment by xacrimon** [comment]
@kchibisov A similar feature is merged now anyway but we are sort of forced to do this due to Metal disabling GC for nonvisible windows on aarch OSX. There isn't a better way we know of to do it on that platform.

**Pull request closed without merging**
